File: modules/angular2/src/core/linker/view_pool.ts
import {Inject, Injectable, OpaqueToken} from 'angular2/src/core/di';

import {isPresent, isBlank, CONST_EXPR} from 'angular2/src/facade/lang';
import {MapWrapper, Map} from 'angular2/src/facade/collection';

import * as viewModule from './view';

export const APP_VIEW_POOL_CAPACITY = CONST_EXPR(new OpaqueToken('AppViewPool.viewPoolCapacity'));

@Injectable()
export class AppViewPool {
  /** @internal */
  _poolCapacityPerProtoView: number;
  /** @internal */
  _pooledViewsPerProtoView = new Map<viewModule.AppProtoView, Array<viewModule.AppView>>();

  constructor(@Inject(APP_VIEW_POOL_CAPACITY) poolCapacityPerProtoView) {
    this._poolCapacityPerProtoView = poolCapacityPerProtoView;
  }

  getView(protoView: viewModule.AppProtoView): viewModule.AppView {
    var pooledViews = this._pooledViewsPerProtoView.get(protoView);
    if (isPresent(pooledViews) && pooledViews.length > 0) {
      return pooledViews.pop();
    }
    return null;
  }

  returnView(view: viewModule.AppView): boolean {
    var protoView = view.proto;
    var pooledViews = this._pooledViewsPerProtoView.get(protoView);
    if (isBlank(pooledViews)) {
      pooledViews = [];
      this._pooledViewsPerProtoView.set(protoView, pooledViews);
    }
    var haveRemainingCapacity = pooledViews.length < this._poolCapacityPerProtoView;
    if (haveRemainingCapacity) {
      pooledViews.push(view);
    }
    return haveRemainingCapacity;
  }
}

File: lib/src/core/linker/view_pool.dart
library angular2.src.core.linker.view_pool;

import "package:angular2/src/core/di.dart" show Inject, Injectable, OpaqueToken;
import "package:angular2/src/facade/lang.dart" show isPresent, isBlank;
import "package:angular2/src/facade/collection.dart" show MapWrapper, Map;
import "view.dart" as viewModule;

const APP_VIEW_POOL_CAPACITY =
    const OpaqueToken("AppViewPool.viewPoolCapacity");

@Injectable()
class AppViewPool {
  /** @internal */
  num _poolCapacityPerProtoView;
  /** @internal */
  var _pooledViewsPerProtoView =
      new Map<viewModule.AppProtoView, List<viewModule.AppView>>();
  AppViewPool(@Inject(APP_VIEW_POOL_CAPACITY) poolCapacityPerProtoView) {
    this._poolCapacityPerProtoView = poolCapacityPerProtoView;
  }
  viewModule.AppView getView(viewModule.AppProtoView protoView) {
    var pooledViews = this._pooledViewsPerProtoView[protoView];
    if (isPresent(pooledViews) && pooledViews.length > 0) {
      return pooledViews.removeLast();
    }
    return null;
  }

  bool returnView(viewModule.AppView view) {
    var protoView = view.proto;
    var pooledViews = this._pooledViewsPerProtoView[protoView];
    if (isBlank(pooledViews)) {
      pooledViews = [];
      this._pooledViewsPerProtoView[protoView] = pooledViews;
    }
    var haveRemainingCapacity =
        pooledViews.length < this._poolCapacityPerProtoView;
    if (haveRemainingCapacity) {
      pooledViews.add(view);
    }
    return haveRemainingCapacity;
  }
}
